src/commands/crater.rs

```diff
@@ -0,0 +1,187 @@
+
+#[derive(Parser, Debug, Default)]
+#[clap(about = "Pre-instantiate all apps in a repository")]
+pub struct CraterCommand {
+    /// Directory containing bindle repository.
+    #[clap(env = "BINDLE_DIRECTORY")]
+    pub directory: PathBuf,
+}
+
+impl CraterCommand {
+    /// Iterate over all the apps in the specified Bindle repository, attempt to pre-instantiate all the components
+    /// of each app, and print the result of each attempt to stdout.
+    pub async fn run(self) -> Result<()> {
+        // Note that we access the Bindle repository directly instead of going through a Bindle server since the
+        // latter is prohibitively slow (especially for large invoices).  That means duplicating some of the code
+        // in `spin-loader`, unfortunately, but it also allows us to skip copying static assets, etc. since they're
+        // not relevant.
+
+        println!("loading index...");
+        let directory = Rc::from(self.directory.as_ref());
+        let index = StrictEngine::default();
+        FileProvider::new(&directory, index.clone()).await;
+        println!("index loaded; getting invoices...");
+
+        let mut invoices = Vec::new();
+        {
+            let mut offset = 0;
+            loop {
+                let matches = index
+                    .query(
+                        "",
+                        "",
+                        SearchOptions {
+                            offset,
+                            limit: 100,
+                            strict: true,
+                            yanked: false,
+                        },
+                    )
+                    .await?;
+
+                if matches.invoices.is_empty() {
+                    break;
+                } else {
+                    offset += u64::try_from(matches.invoices.len()).unwrap();
+                    invoices.extend(matches.invoices);
+                }
+            }
+        }
+
+        let total = invoices.len();
+        println!("got {total} invoices");
+
+        let mut stream = stream::iter(invoices)
+            .enumerate()
+            .map(|(index, invoice)| {
+                let directory = directory.clone();
+
+                async move {
+                    (
+                        index + 1,
+                        pre_instantiate(&directory, &invoice).await,
+                        invoice.bindle.id,
+                    )
+                }
+            })
+            .buffer_unordered(MAX_CONCURRENCY);
+
+        // Note that these will print somewhat out-of-order with respect to `index` due to the `buffer_unordered`
+        // call above.  If that becomes annoying, we can either stop using `buffer_unordered` or accumulate the
+        // results and sort them before printing.
+        while let Some((index, result, id)) = stream.next().await {
+            print!("({index:>5} of {total}) app {id}: ");
+            if let Err(e) = result {
+                println!("failed: {e:?}");
```

Comment:
The reason I'm using `stdout` here is that it's reporting a failure in one of the apps being tested rather than a failure of `spin-crater` itself.  The idea is that we can run `spin crater some_directory > results.txt` and only write to `stderr` if `spin-crater` itself has an issue which prevents it from running normally (e.g. if the specified directory doesn't exist).